proof/refine/ARM/Ipc_R.thy

```diff
@@ -2146,130 +2146,144 @@ lemma replyPush_corres:
-                   dest: has_reply_cap_cte_wpD
-                  dest!: valid_reply_caps_awaiting_reply cte_wp_at_is_reply_cap_toI)
-  apply (case_tac state, simp_all add: bind_assoc)
-  apply (simp add: isReply_def liftM_def)
-  apply (rule corres_symb_exec_r[OF _ getCTE_sp getCTE_inv, rotated])
-   apply (rule no_fail_pre, wp)
-   apply clarsimp
-  apply (rename_tac mdbnode)
-  apply (rule_tac P="Q" and Q="Q" and P'="Q'" and Q'="(\<lambda>s. Q' s \<and> R' s)" for Q Q' R'
-            in stronger_corres_guard_imp[rotated])
-    apply assumption
-   apply (rule conjI, assumption)
-   apply (clarsimp simp: cte_wp_at_ctes_of)
-   apply (drule cte_wp_at_is_reply_cap_toI)
-   apply (erule(4) reply_cap_end_mdb_chain)
-  apply (rule corres_assert_assume[rotated], simp)
-  apply (simp add: getSlotCap_def)
-  apply (rule corres_symb_exec_r[OF _ getCTE_sp getCTE_inv, rotated])
-   apply (rule no_fail_pre, wp)
-   apply (clarsimp simp: cte_wp_at_ctes_of)
-  apply (rule corres_assert_assume[rotated])
-   apply (clarsimp simp: cte_wp_at_ctes_of)
-  apply (rule corres_guard_imp)
-    apply (rule corres_split_deprecated [OF _ threadget_fault_corres])
-      apply (case_tac rv, simp_all add: fault_rel_optionation_def bind_assoc)[1]
-       apply (rule corres_split_deprecated [OF _ dit_corres])
-         apply (rule corres_split_deprecated [OF _ cap_delete_one_corres])
-           apply (rule corres_split_deprecated [OF _ sts_corres])
-              apply (rule possibleSwitchTo_corres)
-             apply simp
-            apply (wp set_thread_state_runnable_valid_sched set_thread_state_runnable_weak_valid_sched_action sts_st_tcb_at' sts_st_tcb' sts_valid_queues sts_valid_objs' delete_one_tcbDomain_obj_at'
-                   | simp add: valid_tcb_state'_def)+
-        apply (strengthen cte_wp_at_reply_cap_can_fast_finalise)
-        apply (wp hoare_vcg_conj_lift)
-         apply (rule hoare_strengthen_post [OF do_ipc_transfer_non_null_cte_wp_at])
-          prefer 2
-          apply (erule cte_wp_at_weakenE)
-          apply (fastforce)
-         apply (clarsimp simp:is_cap_simps)
-        apply (wp weak_valid_sched_action_lift)+
-       apply (rule_tac Q="\<lambda>_. valid_queues' and valid_objs' and cur_tcb' and tcb_at' receiver and (\<lambda>s. sch_act_wf (ksSchedulerAction s) s)" in hoare_post_imp, simp add: sch_act_wf_weak)
-       apply (wp tcb_in_cur_domain'_lift)
-      defer
-      apply (simp)
-      apply (wp)+
-    apply clarsimp
-    apply (rule conjI, erule invs_valid_objs)
-    apply (rule conjI, clarsimp)+
-    apply (rule conjI)
-     apply (erule cte_wp_at_weakenE)
-     apply clarsimp
-     apply (rule conjI, rule refl)
-     apply (fastforce)
-    apply (clarsimp simp: invs_def valid_sched_def valid_sched_action_def)
-   apply (simp)
-   apply (auto simp: invs'_def valid_state'_def)[1]
+crunches handle_fault_reply
+  for pspace_aligned[wp]: pspace_aligned
+  and pspace_distinct[wp]: pspace_distinct
```

Comment:
I would like to move these up to `Move.thy` but they take advantage of unfolding arch-specific functions and so they can't be moved trivially.